Objects/enumobject.c

```diff
@@ -431,16 +432,20 @@ reversed_next(reversedobject *ro)
     PyObject *item;
     Py_ssize_t index = ro->index;
 
+    Py_BEGIN_CRITICAL_SECTION(ro);
     if (index >= 0) {
         item = PySequence_GetItem(ro->seq, index);
         if (item != NULL) {
             ro->index--;
```

Comment:
We don't declare the critical section for the list_next
https://github.com/python/cpython/blob/dacc5ac71a8e546f9ef76805827cb50d4d40cabf/Objects/listobject.c#L3890-L3911

Why do we have to declare the critical section here?
And what about just handling as the out-of-bound, 
and if the ro->index needs to be decremented, why not just use atomic operation in here?

cc @colesbury

Reply:
The check `index >= 0` and decrement `ro->index--` need to happen atomically. We could use the atomic `_Py_atomic_compare_exchange` for this, but it would increase code complexity. If performance is a concern, I can update the PR with this approach. The  `PySequence_GetItem` is not inside the critical section, but I assumed it to be (or to be made later) thread safe